xds/src/main/java/io/grpc/xds/PriorityLoadBalancer.java

```diff
@@ -0,0 +1,297 @@
+      String priority = priorities.get(i);
+      if (!children.containsKey(priority)) {
+        ChildLbState child = new ChildLbState(priority);
+        children.put(priority, child);
+        child.updateResolvedAddresses();
+        return; // Give priority i time to connect.
+      }
+      ChildLbState child = children.get(priority);
+      child.reactivate();
+      if (child.connectivityState.equals(READY) || child.connectivityState.equals(IDLE)) {
+        logger.log(XdsLogLevel.DEBUG, "Shifted to priority {0}", priority);
+        currentPriority = priority;
+        updateOverallState(child.connectivityState, child.picker);
+        for (int j = i + 1; j < priorities.size(); j++) {
+          String p = priorities.get(j);
+          if (children.containsKey(p)) {
+            children.get(p).deactivate();
+          }
+        }
+        return;
+      }
+      if (child.failOverTimer != null && child.failOverTimer.isPending()) {
+        return; // Give priority i time to connect.
+      }
+    }
+    // TODO(zdapeng): Include error details of each priority.
+    logger.log(XdsLogLevel.DEBUG, "All priority failed");
+    updateOverallState(TRANSIENT_FAILURE, new ErrorPicker(Status.UNAVAILABLE));
+  }
+
+  private void updateOverallState(ConnectivityState state, SubchannelPicker picker) {
+    if (!state.equals(currentConnectivityState) || !picker.equals(currentPicker)) {
+      currentConnectivityState = state;
+      currentPicker = picker;
+      helper.updateBalancingState(state, picker);
+    }
+  }
+
+  private final class ChildLbState {
+    final String priority;
+    final ChildHelper childHelper;
+    final GracefulSwitchLoadBalancer lb;
+    // Timer to fail over to the next priority if not connected in 10 sec. Scheduled only once at
+    // child initialization.
+    final ScheduledHandle failOverTimer;
+    // Timer to delay shutdown and deletion of the priority. Scheduled whenever the child is
+    // deactivated.
+    @Nullable ScheduledHandle deletionTimer;
+    @Nullable String policy;
+    ConnectivityState connectivityState = CONNECTING;
+    SubchannelPicker picker = BUFFER_PICKER;
+
+    ChildLbState(final String priority) {
+      this.priority = priority;
+      childHelper = new ChildHelper();
+      lb = new GracefulSwitchLoadBalancer(childHelper);
+
+      class FailOverTask implements Runnable {
+        @Override
+        public void run() {
+          if (deletionTimer != null && deletionTimer.isPending()) {
+            // The child is deactivated.
+            return;
+          }
+          int idx = priorityToIndex.get(priority);
+          updateOverallState(
+              TRANSIENT_FAILURE,
+              new ErrorPicker(
+                  Status.UNAVAILABLE.withDescription(
+                      "Resources for priority " + priority + " not reachable in 10 seconds")));
+          logger.log(XdsLogLevel.DEBUG, "Priority {0} failed over to next", priority);
+          tryNextPriority(idx + 1);
```

Comment:
I haven't reviewed all of this code, but I happened to notice this while replying to the comment above, and I thought I should mention that this looks wrong.

When a given child goes into TRANSIENT_FAILURE, we don't necessarily know that we should go to the priority after the current one.  For example, consider the following case:

1. Priority list is [A, B, C], and we are currently using A.
2. We get an update that changes the priority list to [B, A, C].  We keep the current priority set to A, but we immediately start trying to connect to B, since it is now a higher priority.
3. While B is still trying to connect, A fails.

In this case, there is no need to start trying C, because we are already trying B.  We should not try C unless B winds up failing.

To address this, I think that `TryNextPriority()` should always start from the highest priority and work its way down the list, even if it's called because a priority halfway down the list failed.  At worst, this will cause it to iterate over some priorities that are already known to be bad, but the current state of those priorities should be cached, so that shouldn't cause much overhead.

Reply:
> 1 Priority list is [A, B, C], and we are currently using A.

If A is READY, then this timer is cancelled.

> 2. We get an update that changes the priority list to [B, A, C]. We keep the current priority set to A, but we immediately start trying to connect to B, since it is now a higher priority.

If A was not ready yet, A is deactivated because B bumps up, so when the timer expires, it does nothing https://github.com/grpc/grpc-java/pull/7039/files/ac67d3b8049f18aa3c572ea3ad9a2b1375b91f03#diff-73965829139c9a616f79961469364652R183-R184

So in either case, when B bumps up and A fails while B is still trying to connect, A won't fail over to C.

Reply:
But I agree that let TryNextPriority() always start from the highest priority will make the code much more easier to read with less concern. It does not cause overhead, only when debugging it might take more iterations to track the code.

Reply:
> If A was not ready yet, A is deactivated because B bumps up

I don't think A should be deactivated in that case, becuse it is still present in the config, and it is the currently the highest-priority child that is READY.  It should only get deactivated once B becomes READY, at which point we should use the same code-path as if the config had never changed but B had previously been in TRANSIENT_FAILURE (i.e., as if we are moving back to a higher-priority child that has just recovered from a failure).

Reply:
>> If A was not ready yet, A is deactivated because B bumps up

> I don't think A should be deactivated in that case, becuse it is still present in the config, and it is the currently the highest-priority child that is READY. 

The assumption in that case A is not READY. If A had been READY the timer itself would be cancelled. I think that "A should only get deactivated once B becomes READY" is right. Anyway, I will let TryNextPriority() always start from p0. Otherwise there are so many edge cases and subtleties to worry about.